rosbag2_transport/src/rosbag2_transport/player.cpp

```diff
@@ -901,69 +922,125 @@ void PlayerImpl::load_storage_content()
+    if (message_queue_size < queue_lower_boundary) {
+      enqueue_up_to_boundary(queue_upper_boundary, message_queue_size);
     } else {
       lk.unlock();
       std::this_thread::sleep_for(std::chrono::milliseconds(1));
     }
   }
 }
 
-void PlayerImpl::enqueue_up_to_boundary(size_t boundary)
+void PlayerImpl::enqueue_up_to_boundary(const size_t boundary, const size_t message_queue_size)
 {
-  rosbag2_storage::SerializedBagMessageSharedPtr message;
-  for (size_t i = message_queue_.size_approx(); i < boundary; i++) {
-    if (!reader_->has_next()) {
-      break;
+  // Read messages from input bags in a round robin way
+  size_t input_bag_index = 0u;
+  for (size_t i = message_queue_size; i < boundary; i++) {
+    const auto & reader = input_bags_[input_bag_index].first;
+    // We are supposed to have at least one bag with messages to read
+    if (reader->has_next()) {
+      message_queue_.push(reader->read_next());
     }
-    message = reader_->read_next();
-    message_queue_.enqueue(message);
+    input_bag_index = (input_bag_index + 1) % input_bags_.size();
   }
 }
 
 void PlayerImpl::play_messages_from_queue()
 {
-  // Note: We need to use message_queue_.peek() instead of message_queue_.try_dequeue(message)
-  // to support play_next() API logic.
-  rosbag2_storage::SerializedBagMessageSharedPtr message_ptr = peek_next_message_from_queue();
-  { // Notify play_next() that we are ready for playback
-    // Note: We should do notification that we are ready for playback after peeking pointer to
-    // the next message. message_queue_.peek() is not allowed to be called from more than one
-    // thread concurrently.
+  { // Notify play_next()/seek() that we are ready for playback
     std::lock_guard<std::mutex> lk(ready_to_play_from_queue_mutex_);
     is_ready_to_play_from_queue_ = true;
     ready_to_play_from_queue_cv_.notify_all();
   }
-  while (rclcpp::ok() && !stop_playback_ &&
-    message_ptr != nullptr && !shall_stop_at_timestamp(message_ptr->recv_timestamp))
-  {
-    // Do not move on until sleep_until returns true
-    // It will always sleep, so this is not a tight busy loop on pause
-    while (rclcpp::ok() && !clock_->sleep_until(message_ptr->recv_timestamp)) {
-      if (std::atomic_exchange(&cancel_wait_for_next_message_, false)) {
-        break;
+  rosbag2_storage::SerializedBagMessageSharedPtr message_ptr = take_next_message_from_queue();
+
+  // While we haven't stopped playing
+  while (rclcpp::ok() && !stop_playback_) {
+    // If we had run out of messages before but are starting to play next again, e.g., after a
+    // seek()
+    if (play_next_.load()) {
+      std::lock_guard<std::mutex> lk(skip_message_in_main_play_loop_mutex_);
+      skip_message_in_main_play_loop_ = false;
+      cancel_wait_for_next_message_ = false;
+      // Only take again if we need
+      if (nullptr == message_ptr) {
+        message_ptr = take_next_message_from_queue();
```

Comment:
I think it would be better to move this block to the end of the loop to have more context as to why we need it.

Reply:
great idea, thank you. I believe I can then remove the `if (nullptr == message_ptr)` check then, because that was meant to guard against taking again if we just took a message at the beginning of `play_messages_from_queue()`, before the `while`s. Changed in c6d2ef5951782a6405fc9fbe2d71e769cec7cac6